cpp/src/arrow/compute/kernels/compare.cc

```diff
@@ -152,21 +216,19 @@ Status Compare(FunctionContext* context, const Datum& left, const Datum& right,
                struct CompareOptions options, Datum* out) {
   DCHECK(out);
 
-  DCHECK_EQ(left.kind(), Datum::ARRAY);
-  DCHECK_EQ(right.kind(), Datum::SCALAR);
-  DCHECK(left.type()->Equals(right.type()));
-
-  auto array = left.make_array();
-  auto type = array->type();
-
+  auto type = left.type();
+  DCHECK(type->Equals(right.type()));
+  // Requires that both types are equal.
   auto fn = MakeCompareFilterFunction(context, *type, options);
   if (fn == nullptr) {
     return Status::NotImplemented("Compare not implemented for type ", type->ToString());
   }
 
   FilterBinaryKernel filter_kernel(fn);
   detail::PrimitiveAllocatingBinaryKernel kernel(&filter_kernel);
-  out->value = ArrayData::Make(filter_kernel.out_type(), array->length());
+
+  const int64_t length = FilterBinaryKernel::out_length(left, right);
+  out->value = ArrayData::Make(filter_kernel.out_type(), length);
 
   return kernel.Call(context, left, right, out);
```

Comment:
not this change, but I think the convention we should be using for now is that we use detail::Invoke to handle chunked arrays in the top level method, and worry about the parallelization when we expose the raw kernels?  Thoughts?

Reply:
I agree but I suspect it will be hard to make with the current Invoke method, as the shape can vary, e.g. the signature is `(S,T) -> R`, where S,T,R can be Array, ChunkedArray, Scalar.

Reply:
Hmm, so I think the complication is scalar. (I thought invoke handled combinations of Array and ChunkedArray).  One way of potentially doing this is converting the Kernel into a Unary one when the function is called with scalar and use the InvokeUnary instead (and handle scalar to scalar here).